components/ai_chat/content/browser/ai_chat_brave_search_throttle.cc

```diff
@@ -0,0 +1,159 @@
+#include "brave/components/ai_chat/core/browser/utils.h"
+#include "brave/components/ai_chat/core/common/utils.h"
+#include "components/user_prefs/user_prefs.h"
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/navigation_handle.h"
+#include "content/public/browser/permission_controller.h"
+#include "content/public/browser/permission_request_description.h"
+#include "content/public/browser/web_contents.h"
+
+namespace ai_chat {
+
+// static
+std::unique_ptr<AIChatBraveSearchThrottle>
+AIChatBraveSearchThrottle::MaybeCreateThrottleFor(
+    std::unique_ptr<Delegate> delegate,
+    content::NavigationHandle* navigation_handle,
+    AIChatService* ai_chat_service) {
+  auto* web_contents = navigation_handle->GetWebContents();
+  if (!web_contents) {
+    return nullptr;
+  }
+
+  if (!delegate || !ai_chat_service ||
+      !ai_chat::IsAIChatEnabled(
+          user_prefs::UserPrefs::Get(web_contents->GetBrowserContext())) ||
+      !::ai_chat::IsOpenLeoButtonFromBraveSearchURL(
+          navigation_handle->GetURL())) {
+    return nullptr;
+  }
+
+  return std::make_unique<AIChatBraveSearchThrottle>(
+      std::move(delegate), navigation_handle, ai_chat_service);
+}
+
+AIChatBraveSearchThrottle::AIChatBraveSearchThrottle(
+    std::unique_ptr<Delegate> delegate,
+    content::NavigationHandle* handle,
+    AIChatService* ai_chat_service)
+    : content::NavigationThrottle(handle),
+      delegate_(std::move(delegate)),
+      ai_chat_service_(ai_chat_service) {
+  CHECK(delegate_);
+  CHECK(ai_chat_service_);
+}
+
+AIChatBraveSearchThrottle::~AIChatBraveSearchThrottle() = default;
+
+AIChatBraveSearchThrottle::ThrottleCheckResult
+AIChatBraveSearchThrottle::WillStartRequest() {
+  content::WebContents* web_contents = navigation_handle()->GetWebContents();
+  if (!web_contents || !navigation_handle()->IsInPrimaryMainFrame() ||
+      !IsOpenLeoButtonFromBraveSearchURL(navigation_handle()->GetURL()) ||
+      !IsBraveSearchURL(web_contents->GetLastCommittedURL())) {
+    // Uninterested navigation for this throttle.
+    return content::NavigationThrottle::PROCEED;
+  }
+
+  // Check if nonce in HTML tag matches the one in the URL.
+  AIChatTabHelper* helper =
+      ai_chat::AIChatTabHelper::FromWebContents(web_contents);
+  if (!helper) {
+    return content::NavigationThrottle::CANCEL;
+  }
+
+  helper->ValidateOpenLeoButtonNonce(
+      navigation_handle()->GetURL().ref(),
+      base::BindOnce(&AIChatBraveSearchThrottle::OnNonceValidationResult,
+                     weak_factory_.GetWeakPtr()));
+  return content::NavigationThrottle::DEFER;
+}
+
+void AIChatBraveSearchThrottle::OpenLeoWithStagedConversions() {
+  content::WebContents* web_contents = navigation_handle()->GetWebContents();
+  if (!web_contents) {
+    return;
+  }
+
+  ai_chat::AIChatTabHelper* helper =
+      ai_chat::AIChatTabHelper::FromWebContents(web_contents);
+  if (!helper) {
+    return;
+  }
+
+  ConversationHandler* conversation =
```

Comment:
It seems like this should be encapsulated somewhere else, preferably in AIChatService. `AIChatBraveSearchThrottle` shouldn't have to know so much about the details of what is going on internally and this is also not easily unit testable. So it would be `AIChatService::OpenLeoWithStagedConversations` that takes the helper weakptr and the delegate `OpenLeo` method which really should be static or not a class at all because it doesn't contain any state and then there are no concerns with UAF

Reply:
to be clear I'm talking about everything from this line to the end of the method

Reply:
it seems that @cdesouza-chromium already came to a similar conclusion about the delegate https://github.com/brave/brave-core/pull/26082/files#r1812640057

Reply:
Encapsulate to AIChatService::OpenLeoWithStagedConversations and always check the permission in there.
I didn't check the permission in the actual open Leo sidebar panel util method as IMO that can be reused for browser initiated requests, for example, open sidebar via context menu.